packages/authorization-updater/src/readModelService.ts

```diff
@@ -1,82 +1,127 @@
-    );
+      if (data) {
+        const result = EService.safeParse(data.data);
 
-    if (data) {
-      const result = EService.safeParse(data.data);
+        if (!result.success) {
+          throw genericInternalError(
+            `Unable to parse eservices item: result ${JSON.stringify(
+              result
+            )} - data ${JSON.stringify(data)} `
+          );
+        }
+
+        return result.data;
+      }
+
+      return undefined;
+    },
+
+    async getClientsIdFromPurpose(purposeId: PurposeId): Promise<ClientId[]> {
+      const data = await clients
+        .find(
+          {
+            $or: [
+              {
+                "data.purposes.purpose.purposeId": purposeId,
+              },
+              {
+                "data.purposes": purposeId,
+              },
+            ],
+          },
+          { projection: { data: true } }
+        )
+        .map((c) => c.data)
+        .toArray();
+
+      const result = z.array(z.object({ id: z.string() })).safeParse(data);
 
       if (!result.success) {
         throw genericInternalError(
-          `Unable to parse eservices item: result ${JSON.stringify(
+          `Unable to parse clients item: result ${JSON.stringify(
             result
           )} - data ${JSON.stringify(data)} `
         );
       }
 
-      return result.data;
-    }
-
-    return undefined;
-  }
-
-  async function getClientsIdFromPurpose(
-    purposeId: PurposeId
-  ): Promise<ClientId[]> {
-    const data = await clients
-      .find(
-        {
-          $or: [
-            {
-              "data.purposes.purpose.purposeId": purposeId,
-            },
-            {
-              "data.purposes": purposeId,
-            },
-          ],
-        },
+      return result.data.map((c) => unsafeBrandId<ClientId>(c.id));
+    },
+
+    async getPurposeById(id: PurposeId): Promise<Purpose | undefined> {
+      const data = await purposes.findOne(
+        { "data.id": id },
         { projection: { data: true } }
-      )
-      .map((c) => c.data)
-      .toArray();
+      );
+
+      if (data) {
+        const result = Purpose.safeParse(data.data);
+
+        if (!result.success) {
+          throw genericInternalError(
+            `Unable to parse purpose item: result ${JSON.stringify(
+              result
+            )} - data ${JSON.stringify(data)} `
+          );
+        }
 
-    const result = z.array(z.object({ id: z.string() })).safeParse(data);
+        return result.data;
+      }
+
+      return undefined;
+    },
 
-    if (!result.success) {
-      throw genericInternalError(
-        `Unable to parse clients item: result ${JSON.stringify(
-          result
-        )} - data ${JSON.stringify(data)} `
+    async getAgreement(
+      eserviceId: EServiceId,
+      consumerId: TenantId
+    ): Promise<Agreement | undefined> {
+      const data = await agreements.findOne(
+        { "data.eserviceId": eserviceId, "data.consumerId": consumerId },
```

Comment:
multiple agreements shares the same tuple (`eserviceId`, `consumerId`), for example there can be one active and several archived (after upgrade).
Generally, in order to obtain the specific agreement we would need to take into consideration also the state, which can be `Active` or `Suspended`.

But this is a consumer, which means that it can potentially run on older messages and that they can refer to agreements that in the time of process have changed their state (and be all `Archived`, for example).

So, in this case I suggest to just take the latest agreement, whose state is `Active`, `Suspended` or `Archived`.

Reply:
Updated